plugins/scale/wayfire/plugins/scale-transform.hpp

```diff
@@ -0,0 +1,190 @@
+#ifndef SCALE_TRANSFORM_HPP
+#define SCALE_TRANSFORM_HPP
+
+#include <wayfire/view-transform.hpp>
+#include <wayfire/nonstd/observer_ptr.h>
+#include <string>
+#include <list>
+#include <algorithm>
+
+namespace wf
+{
+/**
+ * transformer used by scale -- it is an extension of the 2D transformer
+ * with the ability to add overlays
+ */
+
+class scale_transformer;
+/**
+ * Overlay function. It is called after the transformed view was already
+ * rendered.
+ *
+ * @param fb     The framebuffer to render to.
+ * @param damage Damaged region to render.
+ * @param tr     Reference to the transformer calling this function.
+ *               It is not marked const to be able to call the transform
+ *               functions, but this function shuld not modify the
+ *               transform parameters (e.g. scale or alpha).
+ * @param data   Any custom data stored when adding this overlay.
+ */
+using scale_view_overlay = std::function<void (const wf::framebuffer_t& fb,
+    const wf::region_t& damage, scale_transformer& tr,
+    nonstd::observer_ptr<custom_data_t> data)>;
+
+class scale_transformer : public wf::view_2D
+{
+  public:
+    scale_transformer(wayfire_view view) : wf::view_2D(view)
+    {}
+    ~scale_transformer()
+    {}
+
+    struct padding
+    {
+        unsigned int top    = 0;
+        unsigned int left   = 0;
+        unsigned int bottom = 0;
+        unsigned int right  = 0;
+    };
+
+    uint32_t get_z_order() override
+    {
+        return wf::TRANSFORMER_HIGHLEVEL + 1;
+    }
+
+    /* render the transformed view and then add all overlays */
+    void render_with_damage(wf::texture_t src_tex, wlr_box src_box,
+        const wf::region_t& damage, const wf::framebuffer_t& target_fb) override
+    {
+        /* render the transformed view first */
+        view_transformer_t::render_with_damage(src_tex, src_box, damage, target_fb);
+
+        /* call all transformers */
+        for (auto& ol : overlays)
+        {
+            (*ol->ol)(target_fb, damage, *this, ol->data.get());
+        }
+    }
+
+    /**
+     * Add a new overlay that is rendered after this transform.
+     *
+     * @param ol      The callback function that does the rendering.
+     * @param z_order Relative order; overlays are called in order.
+     * @param data    Any custom data to store. It is passed to the
+     *                callback function and is automatically destroyed
+     *                when the overlay is removed.
+     * @param pad     Any extra padding to use. Paddings are not cumulative,
+     *                i.e. if two overlays add a top padding, the larger
+     *                of the two values is used.
+     */
+    void add_overlay(scale_view_overlay *ol, int z_order,
+        std::unique_ptr<custom_data_t> data, const padding& pad = {0, 0, 0, 0})
```

Comment:
I'm putting the stored callback into an std::shared_ptr. This way, it can store a lambda created on the fly, but also a plugin can add the same callback object to multiple transformers (if it does not need to store extra data with it). If this is too complicated, I could just use a unique_ptr, requiring a new callback for each transformer.